### ISSUE TITLE
DataGrid support cross domain data stores

### ISSUE THREAD

**Issue by billreed63** [maqetta/maqetta#469]
Cross domain options for URL:
1) XhrIframeProxy - http://shaneosullivan.wordpress.com/2006/08/01/cross-domain-ajax-requests-in-dojo-and-there-was-much-rejoicing/
This is not a good option as it requires that the server hosting the data source deploy the xip_server.html to support iframe proxy url data passing.

2) window.postMessage - This also require that the server hosting the service add code that will postMessages when the data is loaded in an iframe.

3) CORS - This requires that the server hosting the data source support cross-site access control.

4) dojo.io.script - requires that the data store format be JSONP or javascript (ex. a = {item: 1, item1: 2} )

**Comment by billreed63** [comment]
The best option would be for the new MVC data stores to support cross domain stores, We need to investigate this in dojo 1.7. If this does not bare fruit then we need to roll our own DataStore that supports JSONP.

Example of our own data store widget:

``` javascript
     dojo.provide("davinci.widget.ItemJsonpReadStore");
     dojo.require("dojo.data.ItemFileReadStore");
     dojo.declare("davinci.widget.ItemJsonpReadStore", dojo.data.ItemFileReadStore, {


    callbackparm: '',

    constructor: function(/* Object */ keywordParameters){
        debugger;
        this.inherited(arguments);
        if (keywordParameters.callbackparm){
            this.callbackparm = keywordParameters.callbackparm;
        } else{
            throw "ItemJsonpReadStore missing required paramater 'callbackparm'.";
        }
    },

    _fetchItems: function(  /* Object */ keywordArgs,
    /* Function */ findCallback,
    /* Function */ errorCallback){
    //  summary:
    //      See dojo.data.util.simpleFetch.fetch()
        debugger;
        var self = this,
        filter = function(requestArgs, arrayOfItems){
        var items = [],
        i, key;
        if(requestArgs.query){
        var value,
            ignoreCase = requestArgs.queryOptions ? requestArgs.queryOptions.ignoreCase : false;

        //See if there are any string values that can be regexp parsed first to avoid multiple regexp gens on the
        //same value for each item examined.  Much more efficient.
        var regexpList = {};
        for(key in requestArgs.query){
            value = requestArgs.query[key];
            if(typeof value === "string"){
                regexpList[key] = dojo.data.util.filter.patternToRegExp(value, ignoreCase);
            }else if(value instanceof RegExp){
                regexpList[key] = value;
            }
        }
        for(i = 0; i < arrayOfItems.length; ++i){
            var match = true;
            var candidateItem = arrayOfItems[i];
            if(candidateItem === null){
                match = false;
            }else{
                for(key in requestArgs.query){
                    value = requestArgs.query[key];
                    if(!self._containsValue(candidateItem, key, value, regexpList[key])){
                        match = false;
                    }
                }
            }
            if(match){
                items.push(candidateItem);
            }
        }
        findCallback(items, requestArgs);
        }else{
        // We want a copy to pass back in case the parent wishes to sort the array.
        // We shouldn't allow resort of the internal list, so that multiple callers
        // can get lists and sort without affecting each other.  We also need to
        // filter out any null values that have been left as a result of deleteItem()
        // calls in ItemFileWriteStore.
        for(i = 0; i < arrayOfItems.length; ++i){
            var item = arrayOfItems[i];
            if(item !== null){
                items.push(item);
            }
        }
        findCallback(items, requestArgs);
        }
        };

        if(this._loadFinished){
            filter(keywordArgs, this._getItemsArray(keywordArgs.queryOptions));
        }else{
            //Do a check on the JsonFileUrl and crosscheck it.
            //If it doesn't match the cross-check, it needs to be updated
            //This allows for either url or _jsonFileUrl to he changed to
            //reset the store load location.  Done this way for backwards
            //compatibility.  People use _jsonFileUrl (even though officially
            //private.
            if(this._jsonFileUrl !== this._ccUrl){
                dojo.deprecated("dojo.data.ItemFileReadStore: ",
                    "To change the url, set the url property of the store," +
                    " not _jsonFileUrl.  _jsonFileUrl support will be removed in 2.0");
                this._ccUrl = this._jsonFileUrl;
                this.url = this._jsonFileUrl;
            }else if(this.url !== this._ccUrl){
                this._jsonFileUrl = this.url;
                this._ccUrl = this.url;
            }

            //See if there was any forced reset of data.
            if(this.data != null){
                this._jsonData = this.data;
                this.data = null;
            }

            if(this._jsonFileUrl){
            //If fetches come in before the loading has finished, but while
            //a load is in progress, we have to defer the fetching to be
            //invoked in the callback.
            if(this._loadInProgress){
                this._queuedFetches.push({args: keywordArgs, filter: filter});
            }else{
                this._loadInProgress = true;
                //
                var self = this;
                var handle = null;
                var myHandler = function(data){
                    if(handle !== null){
                        dojo.disconnect(handle);
                    }

                    //Process the items...
                    //fetchHandler(self._processPicasaData(data), request);
                    try{
                        self._getItemsFromLoadedData(data);
                        self._loadFinished = true;
                        self._loadInProgress = false;

                        filter(keywordArgs, self._getItemsArray(keywordArgs.queryOptions));
                        self._handleQueuedFetches();
                    }catch(e){
                        self._loadFinished = true;
                        self._loadInProgress = false;
                        errorCallback(e, keywordArgs);
                    }

                };
                var getArgs = {
                        url: self._jsonFileUrl,
                        preventCache: this.urlPreventCache,
                        //content: content,
                        callbackParamName: self.callbackparm,
                        handle: myHandler
                    };
                    var deferred = dojo.io.script.get(getArgs);

                    deferred.addErrback(function(error){
                        dojo.disconnect(handle);
                        errorHandler(error, request);
                    });

                //
                /*var getArgs = {
                        url: self._jsonFileUrl,
                        handleAs: "json-comment-optional",
                        preventCache: this.urlPreventCache,
                        failOk: this.failOk
                    };
                var getHandler = dojo.xhrGet(getArgs);
                getHandler.addCallback(function(data){
                    try{
                        self._getItemsFromLoadedData(data);
                        self._loadFinished = true;
                        self._loadInProgress = false;

                        filter(keywordArgs, self._getItemsArray(keywordArgs.queryOptions));
                        self._handleQueuedFetches();
                    }catch(e){
                        self._loadFinished = true;
                        self._loadInProgress = false;
                        errorCallback(e, keywordArgs);
                    }
                });
                getHandler.addErrback(function(error){
                    self._loadInProgress = false;
                    errorCallback(error, keywordArgs);
                });*/

                //Wire up the cancel to abort of the request
                //This call cancel on the deferred if it hasn't been called
                //yet and then will chain to the simple abort of the
                //simpleFetch keywordArgs
                var oldAbort = null;
                if(keywordArgs.abort){
                    oldAbort = keywordArgs.abort;
                }
                keywordArgs.abort = function(){
                    var df = getHandler;
                    if(df && df.fired === -1){
                        df.cancel();
                        df = null;
                    }
                    if(oldAbort){
                        oldAbort.call(keywordArgs);
                    }
                };
            }
            }else if(this._jsonData){
            try{
                this._loadFinished = true;
                this._getItemsFromLoadedData(this._jsonData);
                this._jsonData = null;
                filter(keywordArgs, this._getItemsArray(keywordArgs.queryOptions));
            }catch(e){
                errorCallback(e, keywordArgs);
            }
            }else{
                errorCallback(new Error("dojo.data.ItemFileReadStore: No JSON source data was provided as either URL or a nested Javascript object."), keywordArgs);
            }
        }
    }




    });
```

Example of html using our data store widget:

``` html
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN"
        "http://www.w3.org/TR/html4/strict.dtd">
<html>
<head>
    <title>Dojo Visual Loader Test</title>
    <style type="text/css">
        @import "../../../dojo/resources/dojo.css";
        @import "../../../dijit/themes/tundra/tundra.css";
        @import "../../../dijit/themes/dijit.css";
        @import "../../../dijit/tests/css/dijitTests.css"; 

        .oddRow { background-color: #f2f5f9; }
        .population { text-align: right; }
    </style>

    <script type="text/javascript" src="../../../dojo/dojo.js" 
        djConfig="isDebug: false, parseOnLoad: true"></script>
    <script type="text/javascript">
        dojo.require("dijit.dijit");
        dojo.require("dojo.parser");
        dojo.require("dijit.Declaration");
        dojo.require("dojo.data.ItemFileReadStore");
        dojo.require("dojox.data.FlickrStore");
        dojo.require("davinci.widget.ItemJsonpReadStore");
    </script>
</head>
<body class="tundra">
<!--    <span dojoType="dojo.data.ItemFileReadStore" 
        jsId="continentStore"
        url="../../../dijit/tests/_data/countries.json"></span> -->
    <span dojoType="davinci.widget.ItemJsonpReadStore" 
        jsId="continentStore"
        url="http://testjson/test2/SampleData" callbackparm="callback"></span>
    <span dojoType="dojox.data.FlickrStore" jsId="flickrStore"></span>


    <h1 class="testTitle">Dojox Data Demo Table</h1>

    <table dojoType="dijit.Declaration" 
        widgetClass="demo.Table" class="dojoTabular" 
        defaults="{ store: null, query: { query: { name: '*' } }, columns: [ { name: 'Name', attribute: 'name' } ] }">
        <thead dojoAttachPoint="head">
            <tr dojoAttachPoint="headRow"></tr>
        </thead>
        <tbody dojoAttachPoint="body">
            <tr dojoAttachPoint="row">
            </tr>
        </tbody>

        <script type="dojo/method">
            dojo.forEach(this.columns, function(item, idx){
                var icn = item.className||"";
                // add a header for each column
                var tth = document.createElement("th");
                tth.innerHTML = item.name;
                tth.className = icn;
                dojo.connect(tth, "onclick", dojo.hitch(this, "onSort", idx));
                this.headRow.appendChild(tth);

                // and fill in the column cell in the template row
                this.row.appendChild(document.createElement("td"));
                this.row.lastChild.className = icn;
            }, this);
            this.runQuery();
        </script>
        <script type="dojo/method" event="onSort" args="index">
            var ca = this.columns[index].attribute;
            var qs = this.query.sort;
            // clobber an existing sort arrow
            dojo.query("> th", this.headRow).style("background", "").style("paddingRight", "");
            if(qs && qs[0].attribute == ca){
                qs[0].descending = !qs[0].descending;
            }else{
                this.query.sort = [{
                    attribute: ca,
                    descending: false
                }];
            }
            var th = dojo.query("> th", this.headRow)[index];
            th.style.paddingRight = "16px"; // space for the sort arrow
            th.style.background = "url(\""+dojo.moduleUrl("dijit", "themes/tundra/images/arrow"+(this.query.sort[0].descending ? "Up" : "Down")+((dojo.isIE == 6) ? ".gif" : ".png")) + "\") no-repeat 98% 4px";
            this.runQuery();
        </script>
        <script type="dojo/method" event="runQuery">
            this.query.onBegin = dojo.hitch(this, function(){ dojo.query("tr", this.body).orphan(); });
            this.query.onItem = dojo.hitch(this, "onItem");
            this.query.onComplete = dojo.hitch(this, function(){
                dojo.query("tr:nth-child(odd)", this.body).addClass("oddRow");
                dojo.query("tr:nth-child(even)", this.body).removeClass("oddRow");
            });
            this.store.fetch(this.query);
        </script>
        <script type="dojo/method" event="onItem" args="item">
            var tr = this.row.cloneNode(true);
            dojo.query("td", tr).forEach(function(n, i, a){
                var tc = this.columns[i];
                var tv = this.store.getValue(item, tc.attribute)||"";
                if(tc.format){ tv = tc.format(tv, item, this.store); }
                n.innerHTML = tv;
            }, this);
            this.body.appendChild(tr);
        </script>
    </table>

    <span dojoType="demo.Table" store="continentStore"
        query="{ query: { type: 'country' }, sort: [ { attribute: 'name', descending: true } ]  }" 
        id="foo">
        <script type="dojo/method" event="preamble">
            this.columns = [
                { name: "Name", attribute: "name" },
                {   name:       "Population",
                    attribute:  "population",
                    className:  "population"
                }
            ];
        </script>
    </span>
    <span dojoType="demo.Table" store="continentStore"
        query="{ query: { name: 'A*' } }"></span>
    <span dojoType="demo.Table" store="flickrStore"
        query="{ query: { groupid: '27475260@N00' } }">
        <script type="dojo/method" event="preamble">
            this.columns = [
                {   name: "", attribute: "imageUrlSmall", 
                    format: function(value, item, store){
                        return (value.length) ? "<img src='"+value+"'>" : "";
                    }
                },
                { name: "Title", attribute: "title" }
            ];
        </script>
    </span>
    <span dojoType="demo.Table" store="flickrStore"
        query="{ query: { tags: 'dojotoolkit' } }">
        <script type="dojo/method" event="preamble">
            this.columns = [
                {   name: "", attribute: "imageUrlSmall", 
                    format: function(value, item, store){
                        return (value.length) ? "<img src='"+value+"'>" : "";
                    }
                },
                { name: "Title", attribute: "title" }
            ];
        </script>
    </span>
</body>
</html>
```

**Comment by billreed63** [comment]
@peller @JonFerraiolo 

I have looked through the latest MVC code from dojo and do no see any new DataStores that deal with Jsonp. They do have custom DataStores for dealing with specific web services like flickr, etc... .  @peller it seems odd that dojo does not have a generic ItemReadStore for jsonp, but maybe that just assume everyone will use the JsonRestStore. Which makes me think maybe we should use the JsonRestStore instead of write our own JsonpItemStore. 

Any thoughts?

**Comment by billreed63** [comment]
http://www.ibm.com/developerworks/java/library/wa-jsonreststore/?ca=drs-

**Comment by JonFerraiolo** [comment]
Bill and I have talked about this. So far, two possible solutions, both of which are ugly. (1) Use JsonRestStore, which is ugly because servers need to add support for JsonRest before Maqetta could pull data from them, and my guess is approximately 0% of servers today that support JSON data include JsonRest support. (2) Use JSONP, which is ugly because we would have to include 100+ lines of extra runtime JS within maqetta.js (or whatever) to adapt ItemFile*Store to use JSONP instead of XHR, but my guess is that most servers today that support JSON data also support JSONP. My instincts based on partial understanding are to go with (2) and maybe add (1) in the future. Maybe someone will come up with a better option than either (1) or (2).

**Comment by billreed63** [comment]
So the issue now becomes how do we add this bit of Jsonp code? What we need is to replace the ItemFileReadStore with our own store that supports Jsonp, so that DataGrid is using our store to populate the grid. To accomplish this we need to change the declarative code for the store to:

```
<span dojoType="dojox.data.ItemJsonpReadStore" id="ItemFileReadStore_1" 
     jsId="ItemFileReadStore_1"  url="http://testjsonp.com:8081/test4/TestService"
     jsonpCallback="callback"
>
</span>

```

When we do this, our ItemJsonpReadStore is treated as a widget and thus needs a _oam.json metadata file. I don't think we want to add our custom store to dojox.data but where do we put it?

Is there another option?

**Comment by billreed63** [comment]
This is is dependent on #830

**Comment by billreed63** [comment]
Kris Zyp suggests:

```
     dojox.io.xhrPlugins only use cross-site XHR (with Access-Control-Allow-* headers) if you use the addCrossSiteXhr method,
 but you   can alternately use dojox.io.xhrScriptPlugin to use JSONP as the transport for XHR.
Kris

```

So I prototyped it and it seems to work

``` html
<script type="text/javascript">
        dojo.require("dijit.dijit");
                dojo.require("dojo.io.script");
            dojo.require("dojox.grid.DataGrid");
        dojo.require("dojo.data.ItemFileReadStore");
        dojo.require("dojox.io.xhrScriptPlugin");
        dojo.require("dojo.parser");

         dojox.io.xhrScriptPlugin("http://testjson.com:8081/test4/TestService","callback");
    dojox.io.xhrScriptPlugin("http://testjson.com:8081/test4/TestService2","callback");
    </script>
  <body>
       <span dojoType="dojo.data.ItemFileReadStore" 
        jsId="myStorexhr" url="http://testjson.com:8081/test4/TestService">
    </span>
    <span dojoType="dojo.data.ItemFileReadStore" 
        jsId="myStorexhr2" url="http://testjson.com:8081/test4/TestService2">
    </span>

        <div id="gridrestxhr" dojoType="dojox.grid.DataGrid" elasticView="2" query="{ name: '*' }" rowsPerPage="20"
        store="myStorexhr" structure="layoutCountries" rowSelector="20px">
    </div>
    <div id="gridrestxhr2" dojoType="dojox.grid.DataGrid" elasticView="2" query="{ name: '*' }" rowsPerPage="20"
        store="myStorexhr2" structure="layoutCountries" rowSelector="20px">
    </div>
   </body>

```

So we should be able to replace our ItemJsonpReadStore with an ItemFileReadStore and use the plugin instead.

**Comment by billreed63** [comment]
changed datagrid to use dojox.io.xhrScriptPlugin. I patched two bugs in dojox.io.xhrScriptPlugin dojo 1.7 that is in the users workspace the issues for tracking these issues with dojo 1.7 are #950 and #951

September 2011

**Comment by billreed63** [comment]
I deployed a sample http://maqettaservice.appspot.com/sampleservice?callback=xx on google app engine
